These commands and flags are valid for release 2021.2
Updated: August 19th, 2024
Lady Grant - Wrighton lab

# Installing QIIME 2
Please follow the directions for installing Qiime 2 on the server found here:
https://docs.qiime2.org/2022.11/install/native/

Connect to the server:
```
ssh zenith
password: superhardpassword1234
```

When you first log in you are in your directory, you will need to navigate to your project directory.
The project directory for Kerbel is: Agribiome/Kerbel
Once you have navigated there, you will need to make a directory named "16S"
```
cd ../home/projects/Agribiome/Kerbel/
mkdir 16S
```

Command cheat sheet for Mac terminal:  
https://www.makeuseof.com/tag/mac-terminal-commands-cheat-sheet/  
Protocol follows this amplicon pipeline:  
https://github.com/WrightonLabCSU/amplicon_pipeline/blob/main/qiime_tutorial.Rmd  
QIIME 2 home page:  
https://qiime2.org/  
QIIME 2 GitHub:  
https://github.com/qiime2/  

All the original data coming fresh off the sequencer will be put here:
```
/ORG-Data-wrighton/in_house_amplicon_libraries/
```

# 1. Activate QIIME environment on server prior to running anything
Once activated "qiime2-2021.2" will show up before your name. Ex: (qiime2-2021.2) [egrant@zenith 16S]$
If the qiime environment is not activated, the following error will be shown:

```
error "bash: qiime: command not found"
```

This assumes a conda installation and qiime2 version 2021.2

```
source /opt/Miniconda2/miniconda2/bin/activate qiime2-2021.2 
```

# 2. Import sequencing files into QIIME2
This step makes a directory named reads that will contain your barcodes and the forward and reverse primers.  
    A. Make directory for your reads.
```
# mkdir < dir > = make directory < name of directory >
# type "ls" after to make sure you can see your directory
mkdir reads
```
    B. Copy the reads into the reads folder:

Below are two different methods used. The first is a symbolic link and the second is copying. Only use one method.

Method 1: Symbolic links of your reads into this directory will not use as much disk space as copying the reads.   
Both methods are shown below.

R1, R2 and I1 in the file name and correspond to read 1 (forward), read 2 (reverse) and barcodes respectively.
```
# symbolic link method...
ln -s [path to barcodes file] ./barcodes.fastq.gz
ln -s [path to forward reads file]  forward.fastq.gz
ln -s [path to reverse reads file]  reverse.fastq.g
```
Method 2: This step is putting your sequence reads into your newly created directory and renames them to:
< forward.fastq.gz >, < reverse.fastq.gz >, < barcodes.fastq.gz >.

```
# copy method
# cp <file> = copy file
# ./ <directory> = to here
# using the full file path is best here but not needed
# ex: < cp /file path to barcodes and primers ./reads/new name >
cp /ORG-Data-wrighton/in_house_amplicon_libraries/Kerbel_2021_16S/Undetermined_S0_L001_R1_001.fastq.gz ./reads/forward.fastq.gz
cp /ORG-Data-wrighton/in_house_amplicon_libraries/Kerbel_2021_16S/Undetermined_S0_L001_R2_001.fastq.gz ./reads/reverse.fastq.gz
cp /ORG-Data-wrighton/in_house_amplicon_libraries/Kerbel_2021_16S/Undetermined_S0_L001_I1_001.fastq.gz ./reads/barcodes.fastq.gz
```

# 3. Demultiplex your sequences
This step sorts out sequenced reads into separate files for each sample in a sequenced run. If your data is already demultiplexed you may skip to "obtaiining the number of sequences per sample".

⏱ This step takes a while so submit a slurm job for this. (This data set took 2 hours)
** You can also use a screen if slurm is not on your server.
link to screen commands: https://gist.github.com/jctosta/af918e1618682638aa82
```
screen -S <session_name>
# activeate enviornement **new screen needs new environment** 
source /opt/Miniconda2/miniconda2/bin/activate qiime2-2021.2
```
A. Import Sequence files into QIIME 2

The folloing command is for EMP sequences which have one file for forward reads, one file for reverse reads and one file with associated barcodes and is still mulitplexed. Contact your sequencing center if you are unsure of the type.
For more info or if using other data format: https://docs.qiime2.org/2017.12/tutorials/importing/ 
```
qiime tools import --type EMPPairedEndSequences --input-path reads/ --output-path emp-paired-end-sequences.qza
```
Output from command: emp-paired-end-sequences.qza

Before the next step you will need to upload your metadata file as a .txt to the server. It must have the associated barcodes for each sample. Other inforamtion may be useful but is not mandatory. Be sure that there are no duplicates in your sample names.

Before the next step you will need to upload your metadata file as a .txt to the server. It must have the associated barcodes for each sample. Other inforamtion may be useful but is not mandatory. Be sure that there are no duplicates in your sample names.

Lab barcodes are located on Dropbox here:
(https://www.dropbox.com/home/w2-microbiome/Protocols/Wetlab/Sequencing)

  B. Demultiplex the sequences

--i-seqs emp-paired-end-sequences.qza
The "emp-paired-end-sequences.qza" file is the .qza file previouly loaded

--m-barcodes-file metadata.txt
This is your metadata file with your sample ID's and associated barcodes.

--m-barcodes-column barcode-sequence
barcode-sequence indicates that in your mapping file, barcodes are in the ‘barcode-sequence’ column.

--o-per-sample-sequences demux.qza
this outputs a .qza file containing reads separated by sample. If you want to combine reads from multiple sequencing runs, these are the files you would use.

--o-error-correction-details demux-details.qza
Detail about the barcode error corrections

--p-no-rev-comp-mapping-barcodes
specifies barcodes don't need to be reverse complemented but varies based on sequencing center.

--p-no-golay-error-correction
Does NOT perform 12nt Golay error correction on the barcode reads
```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --time=4-00:00:00
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mem=128gb
#SBATCH --mail-user=your.email@colostate.edu
#SBATCH --partition=wrighton-hi,wrighton-low


source /home/opt/Miniconda3/miniconda3/bin/activate qiime2-2021.2

qiime demux emp-paired \
  --m-barcodes-file metadata.txt  \
  --m-barcodes-column BarcodeSequence \
  --p-no-rev-comp-mapping-barcodes \
  --p-no-golay-error-correction\
  --i-seqs emp-paired-end-sequences.qza \
  --o-per-sample-sequences demux.qza \
  --o-error-correction-details demux-details.qza
```
Outputs from step: demux.qza, demux-details.qza

  C. Visualize sequence reads and sequence quality

This command outputs a file containing many data statistics including the number of reads per sample and quality metrics of sequencing. After obtaining the .qzv file, you can visualize it in the QIIME2 view tools: https://view.qiime2.org/ 

```
qiime demux summarize --i-data demux.qza --o-visualization demux.qzv
```
Output from step: demux.qzv

# 4. Run DADA2
A. This command runs dada2 on the data which "denoises" paired-end sequences, dereplicates them, and filters chimeras. This step helps discriminate between true sequence diversity and sequencing errors.

⏱ This step takes a while. (This data set took 8 hours.)

--i-demultiplexed-seqs demux.qza
The .qza file from the previous step

--p-trim-left-f 0
The trim length at the beginning of the forward read

--p-trim-left-r 0
The trim length at the beginning of the reverse read

--p-trunc-len-f 248
The truncation length at the end of the forward read, this number is based off information found in demux.qza

--p-trunc-len-r 248
The truncation length at the end of the reverse read, this number is based off information found in demux.qza

--o-table table.qza
Output table that has the number of each ASV per sample

--o-representative-sequences rep-seqs.qza
A file that contains every unique ASV

--o-denoising-stats denoising-stats.qza
```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --time=4-00:00:00
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mem=20gb
#SBATCH --mail-user=your.email@colostate.edu
#SBATCH --partition=wrighton-hi,wrighton-low

source /home/opt/Miniconda3/miniconda3/bin/activate qiime2-2021.2

qiime dada2 denoise-paired \
  --i-demultiplexed-seqs demux.qza \
  --p-trim-left-f 0 \
  --p-trim-left-r 0 \
  --p-trunc-len-f 248 \
  --p-trunc-len-r 248 \
  --o-table table.qza \
  --o-representative-sequences rep-seqs.qza \
  --o-denoising-stats denoising-stats.qza
```
Outputs from step: table.qza, rep-seqs.qza, denoising-stats.qza

  B. Visualize denoising stats
```
qiime metadata tabulate \
--m-input-file denoising-stats.qza \
--o-visualization denoising-stats.qzv
```
  C. Visualize feature table
```
qiime feature-table summarize \
 --i-table table.qza  \
 --o-visualization table.qzv \
 --m-sample-metadata-file metadata.txt
```
  D. Visualize representative sequences
```
qiime feature-table tabulate-seqs \
--i-data rep-seqs.qza \
--o-visualization rep-seqs.qzv
```
# 5. Taxonomically classify your reads
⏱ This step takes a while. (This data took about 2 hours)  
Classifiers for the Earth Microbiome Primers (EMP) (515f/806R) have been trained and are available on the server for both Silva 138, GTDB 207, and GTDB X. Make sure to check you are using the appropriate version. If you need to train a classifier because of a new database release or you are using different primers, detailed instructions can be found https://docs.qiime2.org/2019.10/tutorials/feature-classifier/

NOTE: Best practices is to classify with both SILVA and GTDB for studies that will involve metagenomics. Also GTDB does not handle pulling out choloroplast or mitochondria very well so we can use SILVA to do so. That step is documented in the Merge tables document. 

Classifying with Silva 138
```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --time=4-00:00:00
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mem=20gb
#SBATCH --mail-user=your.email@colostate.edu
#SBATCH --partition=wrighton-hi,wrighton-low

source /home/opt/Miniconda3/miniconda3/bin/activate qiime2-2021.2

qiime feature-classifier classify-sklearn \
--i-classifier /home/Database/qiime2/classifiers/qiime2-2021.2/silva-138-99-515-806-nb-classifier.qza \
--i-reads /home/projects/Agribiome/Kerbel/16S/16S_Final/rep_seqs/rep-seqs.qza \
--o-classification taxonomy_silva138_16S.qza
```
If using the SILVA classifier above make sure to cite the following sources:

Bokulich, N.A., Robeson, M., Dillon, M.R. bokulich-lab/RESCRIPt. Zenodo. http://doi.org/10.5281/zenodo.3891931

Bokulich, N.A., Kaehler, B.D., Rideout, J.R. et al. Optimizing taxonomic classification of marker-gene amplicon sequences with QIIME 2’s q2-feature-classifier plugin. Microbiome 6, 90 (2018). https://doi.org/10.1186/s40168-018-0470-z
  A. Visualize your taxonomy table
```
qiime metadata tabulate \
--m-input-file taxonomy_silva138_16S.qza \
--o-visualization taxonomy_silva138_16S.qzv
```
  C. Remove Chloroplast, Mitochondria, Unassigned and Eukaryota.
```
qiime taxa filter-table
--i-table table.qza
--i-taxonomy taxonomy.qza
--p-exclude Mitochondria,Chloroplast,Unassigned,Eukaryota
--o-filtered-table table-clean.qza
```

Classifying with GTDB 220
```
#!/bin/bash
#SBATCH --nodes=1
#SBATCH --ntasks=4
#SBATCH --time=4-00:00:00
#SBATCH --mail-type=BEGIN,END,FAIL
#SBATCH --mem=20gb
#SBATCH --mail-user=Eryn.grant@colostate.edu
#SBATCH --partition=wrighton-hi,wrighton-low

source /home/opt/Miniconda3/miniconda3/bin/activate qiime2-2023.9

qiime feature-classifier classify-sklearn \
--i-classifier /home/Database/qiime2/classifiers/qiime2-2023.9/GTDBclassifier220_EMP.qza \
--i-reads /home/projects/Agribiome/Kerbel/16S/16S_Final/rep_seqs/rep-seqs.qza \
--o-classification taxonomy_gtdb_214.qza
```


