# Merging tables, formatting and rairifying samples
## Merging tables and formatting
Before any analysis, we must merge the taxonomy table with the ASV table that was previously created in Qiime 2. There are multiple ways to do this but here we are using left_join.
NOTE: This can also be done in Excel using VLOOKUP, however R will be faster depending on the size of the data set.

```
# Load the needed libraries
library(tidyverse) # ggplot2 + dplyr
library(readxl) # to read excel tables and sheets

# If you are using R projects, setting the directory is not always needed.
# You can use getwd() show you what directory you are working in
#set your working directory
setwd("~/Desktop/Agbiome/Kerbel/01_16S/R_analysis/Kerbel_16S/")
```

Before reading in the tables, open them to check their formatting. Qiime will give a confidence score in a column named "Confidence" on the taxonomy. I remove this in Excel before reading it back in, however this can also be done in R.
Other changes I do in Excel:
Change "Taxon" header to "taxonomy".
Change "Feature ID" to "ASV_ID". (Formatting needed for McToolsR requires that this says "OTU_ID but we will change that later)
Delete: "#q2:types" and "categorical" headers from top of page.

```
# If you are using R projects, setting the directory is not always needed.
# You can use getwd() show you what directory you are working in
#set your working directory
setwd("~/Desktop/Agbiome/Kerbel_paper/01_16S/")

# Read in tables
Taxonomy <- read_xlsx("Merged_taxonomy_Silva138_GTDB220.xlsx", sheet = 1) # sheet name: taxonomy_GTDB_220
ASV_Count <- read_xlsx("Merged_taxonomy_Silva138_GTDB220.xlsx", sheet = 4) # sheet name: CLEAN_counts

# Merge via "ASV_ID"
# Check each data frame to see how R has labeled the columns
ASV_Table_16S <- left_join(ASV_Count, Taxonomy, by = "ASV_ID")
```

Write out table, for more formatting in excel. The first numbered column can be removed. For MCtoolsR, the "taxonomy" column must be on the right end of the file.
The "ASV_ID" column needs a "#" in front of it and needs to say "OTU_ID. Ex: "#OTU_ID". Although we are working with ASV's the formating requires this.

```
write.csv(ASV_Table_16S, "ASV_Table_16S_GTDB220.txt")
```

## Rarefying samples
Here, where we use MCtoolsR and Vegan. There are many different ways to rarefy, but MCtoolsR, which was designed for microbial ecology, has short commands. The data loaded creates list in R that can be easily accessed by "$". We will also use Vegan to collect a few more metrics for the data.
For more information on MCtoolrR:  
https://github.com/leffj/mctoolsr 
```
# Loading taxa tables and metadata using MCtoolsR

# Load needed libraries 
library(mctoolsr)
library(vegan)
library(tidyverse)
set.seed(12345)
# The taxon table itself: "data_loaded"
tax_table_fp = "ASV_Table_16S_GTDB220.txt"
# The metadata: "map_loaded"
map_fp = "I6S_Metadata_R.txt"
                     
# Merge your mapping file and your ASV table
# Common error: Samples with mismatched names will not load
# Common error: NA's in your ASV table will cause blank tables to load
input = load_taxa_table(tax_table_fp, map_fp) # 311 samples loaded
# for merging tables later in ggplot later
map = input$map_loaded %>% 
  rownames_to_column("Sample_ID")
```
This is all modified from Pat Schloss's Code Club: https://riffomonas.org/code_club/2022-04-07-sampling-depth and Dr. Emily Bechtold's notes. Note that in this video he discusses using Good's Coverage as a way to determine depth. DADA2 as implemented in QIIME2 filters out all singletons so Goo'd coverage is not appropriate when using QIIME2 to create the feature table.

Here we pull out the reads per sample to decide what number we want to rarefy at. Then we graph it using a sinple bar graph. Finally the last step is rarfying the reads. The rarefied data set is what will be used for all further analysis. 
```
# read your counts table in
raw <- read.delim("ASV_Table_16S_GTDB220.txt")

# change data to long format. This will have make a column for: 
# ASV's, taxonomy, sample_id and counts
raw_table <- raw %>% 
  pivot_longer(2:318, names_to = "Sample", values_to = "counts")

# calculate number of sequences
sampling_coverage <- raw_table %>% 
  group_by(Sample) %>% 
  summarize(n_seqs = (sum(counts)))
```
```
# density plot
density <- ggplot(sampling_coverage, aes(x=n_seqs)) +
  geom_density()
density
```
[Density plot.pdf](https://github.com/user-attachments/files/16680548/Density.plot.pdf)
```
# histogram
histogram <- ggplot(sampling_coverage, aes(x=n_seqs)) +
  geom_histogram()
histogram
```
[histogram.pdf](https://github.com/user-attachments/files/16680557/histogram.pdf)
```
# scatter plot
sampling_coverage %>%
  ggplot(aes(x=1, y=n_seqs)) +
  geom_jitter() +
  scale_y_log10()
```
[scatter plot.pdf](https://github.com/user-attachments/files/16680561/scatter.plot.pdf)

```
# peek at lowest counts
sampling_coverage %>%
  arrange(n_seqs) %>%
  print(n=50)

#sum of reads
sum(seq_num$read) #8,567,192
```
```
# Returning number of reads per sample
seq_num <- as.data.frame(sort(colSums(input$data_loaded))) %>%
  rownames_to_column("Sample_ID") %>%
  rename(read = `sort(colSums(input$data_loaded))`) %>%
  left_join(map, by = "Sample_ID")
```
```
# Creating a bar graph of reads per sample#, red line on y- intercept shows where we rarified
# Colors = CT = "#3B1C32", MT = "#CA054D", ST = "#6FA37F"
ggplot(seq_num, aes(Sample_ID, read)) + 
  geom_bar(stat="identity", aes(fill = Treatment)) + 
  scale_fill_manual(values = c("#3B1C32","#CA054D", "#6FA37F")) +
  geom_hline(yintercept  = 10000, color = "red") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 5)) + 
  ggtitle("Reads per Sample 16S") + theme(plot.title = element_text(hjust = 0.5)) 
```

```
# Returning number of reads per sample
seq_num <- as.data.frame(sort(colSums(input$data_loaded))) %>%
  rownames_to_column("Sample_ID") %>%
  rename(read = `sort(colSums(input$data_loaded))`) %>%
  left_join(map, by = "Sample_ID")

# Creating a bar graph of reads per sample#, red line on y- intercept shows where we rarified
# Colors = CT = "#3B1C32", MT = "#CA054D", ST = "#6FA37F"
ggplot(seq_num, aes(Sample_ID, read)) + 
  geom_bar(stat="identity", aes(fill = Treatment)) + 
  scale_fill_manual(values = c("#3B1C32","#CA054D", "#6FA37F")) +
  geom_hline(yintercept  = 10000, color = "red") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 5)) + 
  ggtitle("Reads per Sample 16S") + theme(plot.title = element_text(hjust = 0.5)) 

# Rarefying sample reads
set.seed(2000)
input_rar = single_rarefy(input, 10000) #292 Samples remaining
```
![reads_per_sample_16S_12X6](https://github.com/LadyGrant/Kerbel/assets/95941680/84e946a5-1ee4-4665-9222-b2747e142b3b)
